Snowflake.Data.Tests/SFConnectionPoolT.cs

```diff
@@ -504,6 +504,98 @@ public void TestConcurrentConnectionPoolingAsync()
             ConcurrentPoolingAsyncHelper(connStr, true);
         }
 
+        [Test]
+        public void TestRollbackTransactionOnPooledWhenExceptionOccurred()
+        {
+            SnowflakeDbConnectionPool.SetPooling(true);
+            SnowflakeDbConnectionPool.SetMaxPoolSize(10);
+
+            object firstOpenedSession = null;
+            using (var connection = new SnowflakeDbConnection())
+            {
+                connection.ConnectionString = ConnectionString;
+                connection.Open();
+                firstOpenedSession = connection.SfSession.sessionId;
+                connection.BeginTransaction();
+                Assert.Throws<SnowflakeDbException>(() =>
+                {
+                    using (var command = connection.CreateCommand())
+                    {
+                        command.CommandText = "invalid command will throw";
+                        command.ExecuteNonQuery();
+                    }
+                });
+            }
+
+            using (var connectionWithSessionReused = new SnowflakeDbConnection())
+            {
+                connectionWithSessionReused.ConnectionString = ConnectionString;
+                connectionWithSessionReused.Open();
+                Assert.AreEqual(firstOpenedSession, connectionWithSessionReused.SfSession.sessionId);
+                using (var cmd = connectionWithSessionReused.CreateCommand())
+                {
+                    cmd.CommandText = "SELECT CURRENT_TRANSACTION()";
+                    Assert.AreEqual(DBNull.Value, cmd.ExecuteScalar());
+                    Assert.AreEqual(false, connectionWithSessionReused.HasTransactionInProgress());
+                }
+                connectionWithSessionReused.Close();
+            }
+        }
+
+        [Test]
+        public void TestRollbackTransactionOnPooledWhenConnectionClose()
+        {
+            SnowflakeDbConnectionPool.SetPooling(true);
+            SnowflakeDbConnectionPool.SetMaxPoolSize(10);
+            var conn = new SnowflakeDbConnection();
+            conn.ConnectionString = ConnectionString;
+            
+            conn.Open();
+            IDbCommand command = conn.CreateCommand();
+            object firstOpenedSession = conn.SfSession.sessionId;
+            command.CommandText = "BEGIN TRANSACTION ";
+            command.ExecuteNonQuery();
+            conn.Close();
+            
+            conn.Open();
+            Assert.AreEqual(firstOpenedSession, conn.SfSession.sessionId);
+            command.CommandText = "SELECT CURRENT_TRANSACTION()";
+            Assert.AreEqual(DBNull.Value, command.ExecuteScalar());
+            Assert.AreEqual(false, conn.HasTransactionInProgress());
+            conn.Close();
+            
+            Assert.AreEqual(1, SnowflakeDbConnectionPool.GetCurrentPoolSize(), "Should return connection to the pool");
+        }
+
+        private class TestSnowflakeDbFactory : DbProviderFactory 
```

Comment:
Also, I would say, that I have always seen mock of mock, as creating mock which creates mock using library :) Like:
```
factory = Mock<Factory>
command = Mock<Command>
factory.createCommand(return command)
```
WDYT?

Reply:
Mock returning mock is very often considered as antipattern :)

Reply:
Still, it's better than mixing library mocks and custom code mocks IMHO :)

Reply:
It's actually both - factory and DbCommand a spied object because we need to check specific behaviour during integration test :(